plugins/framework-qmk/fu-framework-qmk-device.c

```diff
@@ -0,0 +1,131 @@
+					 req->data,
+					 req->len,
+					 FU_IO_CHANNEL_FLAG_NONE,
+					 error)) {
+		g_prefix_error(error, "failed to write packet: ");
+		return FALSE;
+	}
+
+	/* success */
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static void
+fu_framework_qmk_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DECOMPRESSING, 0, "prepare-fw");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 5, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 15, "reload");
+}
+
+static void
+fu_framework_qmk_device_init(FuFrameworkQmkDevice *self)
+{
+	fu_device_add_icon(FU_DEVICE(self), FU_DEVICE_ICON_INPUT_KEYBOARD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_BCD);
+	fu_device_add_protocol(FU_DEVICE(self), "com.microsoft.uf2");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_ADD_COUNTERPART_GUIDS);
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_REPLUG_MATCH_GUID);
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_RETRY_OPEN);
+	/* revisions indicate incompatible hardware */
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_ADD_INSTANCE_ID_REV);
+	fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_WRITE);
+	fu_device_retry_set_delay(FU_DEVICE(self), 100);
+}
+
+static gboolean
+fu_framework_qmk_device_setup(FuDevice *device, GError **error)
+{
+	g_autoptr(FuHidDescriptor) descriptor = NULL;
+	g_autoptr(FuHidReport) report = NULL;
+	g_autoptr(FuDevice) device_usb = NULL;
+	guint16 version = 0x000;
+
+	descriptor = fu_hidraw_device_parse_descriptor(FU_HIDRAW_DEVICE(device), error);
+	if (descriptor == NULL) {
+		g_prefix_error(error, "Failed to parse descriptor");
+		return FALSE;
+	}
+	report = fu_hid_descriptor_find_report(descriptor,
+					       error,
+					       "usage-page",
+					       FU_FRAMEWORK_QMK_RAW_USAGE_PAGE,
+					       "usage",
+					       FU_FRAMEWORK_QMK_RAW_USAGE_ID,
+					       "collection",
+					       0x01,
+					       NULL);
+	if (report == NULL) {
+		/* Wrong usage page, this hidraw instance is not what we're looking for */
+		return FALSE;
+	}
+
+	device_usb = fu_device_get_backend_parent_with_subsystem(device, "usb:usb_device", NULL);
+	if (device_usb == NULL) {
+		g_prefix_error(error, "USB device NULL");
```

Comment:
it's weird, but g_prefix_error needs a trailing ": " e.g. `g_prefix_error(error, "USB device NULL: ");`